### PR TITLE
feat(cico): New Select Provider Cash-in/cash-out flow

### PR DESCRIPTION
**Pull request by jh2oman**
### A major overhaul of the CICO flow.

1. Adds the new Select Provider Screen which dynamically creates sections for Mobile Money / Credit&Debit / Bank Account based on how many providers return quotes / are available
2. Changes the initial CICO screen to just be selecting currency instead of currency + CICO provider type
3. New 'Spend' Category in the initial CICO Screen
4. No minimum Cash in & Cash out amounts because exchanges can do very low amounts

#### Followup Tickets

* [Analytics events for new CICO flow](https://app.zenhub.com/workspace/o/valora-inc/wallet/issues/2442)
* [E2E tests for Cash-in & Cash-out](https://app.zenhub.com/workspace/o/valora-inc/wallet/issues/2441)

#### A 'Nearly Full' Select Providers Screen
Simulator Screen Shot - iPhone 13 - 2022-05-09 at 12 20 26

#### Cash In Flow

https://user-images.githubusercontent.com/8432644/167495863-b04759a9-7059-4bae-b7cf-8f96af05e721.mp4

#### Cash Out Flow

https://user-images.githubusercontent.com/8432644/167494388-5f09ac04-226d-452d-9511-bf8c82152acd.mp4

### Other changes

The previous implementation actually assumed only one quote per provider and had 'CicoProvider' as the main data structure throughout the flow. This change includes switching to 'Quotes' being the predominant data structure.

### Tested

Unit tests and manual testing. This feature will need to be tested thoroughly in a manual way 

### How others should test

This definitely needs to be tested. All previous Cash in / Cash out functionality should work in this new flow.

### Related issues

https://app.zenhub.com/workspaces/acquisition-squad-sprint-board-6010683afabec1001a090887/issues/valora-inc/wallet/2196

### CODE REVIEW

File: src/app/saga.test.ts [hunk 1]
```diff
@@ -96,16 +99,16 @@ describe('App saga', () => {
   })
 
   it('Handles openScreen deep link with safe origin', async () => {
-    const deepLink = `celo://wallet/openScreen?screen=${Screens.FiatExchangeOptions}&isCashIn=true`
+    const deepLink = `celo://wallet/openScreen?screen=${Screens.FiatExchangeCurrency}&flow=CashIn`
```
Comment (jh2oman) [inline src/app/saga.test.ts:102]: @MuckT Do we have any deeplinks going to this component? I changed the name of the component and the router props so those will break if we have any

File: src/fiatExchanges/FiatExchangeAmount.test.tsx [hunk 2]
```diff
@@ -185,81 +175,10 @@ describe('FiatExchangeAmount cashIn', () => {
     expect(tree.getByTestId('FiatExchangeNextButton')).not.toBeDisabled()
   })
 
-  it('opens a dialog when the cUSD amount is lower than the limit', () => {
```
Comment (jh2oman) [inline src/fiatExchanges/FiatExchangeAmount.test.tsx:188]: We no longer have a 'minimum amount' since exchanges can handle very low amounts.

File: src/fiatExchanges/SelectProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+          <View style={styles.right}>
+            <Text style={styles.linkToOtherScreen}>{t('selectProviderScreen.viewExchanges')}</Text>
+          </View>
+        </View>
+      </Touchable>
+    </View>
+  )
+}
+
+function LocalProviderMobileMoneySection({
```
Comment (jh2oman) [inline src/fiatExchanges/SelectProvider.tsx:225]: These were previously tucked in the Screen where you select your currency. I only realized it existed when digging into the code. Its only used for a few Mobile money providers in the Philippines and Brazil

**Comment by MuckT** [comment]
@jh2oman Running the E2E tests [via dispatch here](https://github.com/valora-inc/wallet/actions/runs/2297208263) as this is not a PR targeting the main branch so they weren't run automatically.

File: src/app/saga.test.ts (hunk 1 above)
Comment (MuckT) [reply src/app/saga.test.ts:102]: @jh2oman - I no active or scheduled campaigns are currently using the `FiatExchangeOptions` deep link. I've also let @laurencedelisle know so we can adjust going forward.

File: src/fiatExchanges/FiatExchangeCurrency.tsx [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+import { Screens } from 'src/navigator/Screens'
+import { StackParamList } from 'src/navigator/types'
+import colors from 'src/styles/colors'
+import fontStyles from 'src/styles/fonts'
+import variables from 'src/styles/variables'
+import { Currency } from 'src/utils/currencies'
+import { CICOFlow, FiatExchangeFlow } from './utils'
+
+type RouteProps = StackScreenProps<StackParamList, Screens.FiatExchangeCurrency>
+type Props = RouteProps
```
Comment (cajubelt) [inline src/fiatExchanges/FiatExchangeCurrency.tsx:24]:
```suggestion
type Props = StackScreenProps<StackParamList, Screens.FiatExchangeCurrency>
```
code golf

File: src/fiatExchanges/FiatExchangeCurrency.tsx [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+  const { flow } = route.params
+  return {
+    ...emptyHeader,
+    headerLeft: () => (
+      <BackButton
+        eventName={FiatExchangeEvents.cico_select_currency_back}
+        eventProperties={{ flow }}
+      />
+    ),
+    headerTitle: i18n.t(`fiatExchangeFlow.${route.params.flow}.selectCurrencyTitle`),
```
Comment (cajubelt) [inline src/fiatExchanges/FiatExchangeCurrency.tsx:40]:
```suggestion
    headerTitle: i18n.t(`fiatExchangeFlow.${flow}.selectCurrencyTitle`),
```
code golf

File: src/fiatExchanges/FiatExchangeCurrency.tsx [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+            body="(cEUR)"
+            selected={selectedCurrency === Currency.Euro}
+            onSelect={onSelectCurrency(Currency.Euro)}
+            containerStyle={{
+              borderTopWidth: 0.5,
+              borderBottomWidth: 0.5,
+            }}
+            testID="radio/cEUR"
+          />
+          <CurrencyRadioItem
```
Comment (cajubelt) [inline src/fiatExchanges/FiatExchangeCurrency.tsx:161]: worth adding cREAL, maybe with a feature flag? We're going to have providers supporting them soon

Comment (cajubelt) [reply src/fiatExchanges/FiatExchangeCurrency.tsx:161]: feel free to punt with an issue at your discretion

**Review by cajubelt — COMMENTED** [review]
reviewed up thru (not including) SelectProvider.test.tsx

Looks pretty good so far!

File: src/fiatExchanges/FiatExchangeCurrency.tsx [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+        currency: selectedCurrency,
+      })
+    }
+    navigate(Screens.FiatExchangeAmount, {
+      currency: selectedCurrency,
+      flow: flow === FiatExchangeFlow.CashIn ? CICOFlow.CashIn : CICOFlow.CashOut,
+    })
+  }
+
+  const onSelectCurrency = (currency: Currency) => () => setSelectedCurrency(currency)
```
Comment (lisaqyli) [inline src/fiatExchanges/FiatExchangeCurrency.tsx:104]: I'm not sure what does this wrapper function do, can just directly use `setSelectedCurrency(currency)` ?

File: src/fiatExchanges/FiatExchangeCurrency.tsx [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+  // const onPressDismissEducationDialog = () => {
+  //   setEducationDialogVisible(false)
+  //   ValoraAnalytics.track(
+  //     isCashIn
+  //       ? FiatExchangeEvents.cico_add_funds_info_cancel
+  //       : FiatExchangeEvents.cico_cash_out_info_cancel
+  //   )
+  // }
+
```
[the remark this answers is not in the document]
Comment (lisaqyli) [reply src/fiatExchanges/FiatExchangeCurrency.tsx:124]: +1

File: src/fiatExchanges/FiatExchangeAmount.tsx [hunk 9]
```diff
@@ -114,10 +114,6 @@ function FiatExchangeAmount({ route }: Props) {
     useCurrencyToLocalAmount(new BigNumber(DOLLAR_ADD_FUNDS_MAX_AMOUNT), Currency.Dollar) ||
     new BigNumber(0)
 
-  const localCurrencyMinAmount =
```
Comment (lisaqyli) [inline src/fiatExchanges/FiatExchangeAmount.tsx:117]: oh I'm unaware we are not doing the min amount any more?

File: locales/base/translation.json [hunk 10]
```diff
@@ -1217,5 +1217,47 @@
...
+    "oneHour": "Less Than 1 Hour",
+    "bestRate": "Best Rate",
+    "header": "Select Payment Method"
+  },
+  "fiatExchangeFlow": {
+    "CashIn": {
+      "fiatExchangeTitle": "Add Funds",
+      "fiatExchangeSubtitle": "Fund your Valora account from a bank, card or exchange.",
+      "selectCurrencyTitle": "Add Funds",
+      "exchangeAmmountTitle": "Add {{currency}}",
```
Comment (lisaqyli) [inline locales/base/translation.json:1247]: there is an extra m in `Amount`, same case for cashout

File: src/fiatExchanges/SelectProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+
+export default function SelectProviderScreen({ route, navigation }: Props) {
+  const dispatch = useDispatch()
+  const userLocation = useSelector(userLocationDataSelector)
+  const account = useSelector(currentAccountSelector)
+  const localCurrency = useSelector(getLocalCurrencyCode)
+  const [noPaymentMethods, setNoPaymentMethods] = useState(false)
+  const { flow } = route.params
+
+  const currencyToBuy = {
```
Comment (lisaqyli) [inline src/fiatExchanges/SelectProvider.tsx:67]: i thought this const also is used in withdraw flow, it's a bit confusing to name it `currencyToBuy`, maybe `currencyInTransaction`

File: src/fiatExchanges/SelectProvider.tsx [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+
+      const localProviders = await fetchLocalCicoProviders()
+      const availableLocalProviders = getAvailableLocalProviders(
+        localProviders,
+        flow,
+        userLocation.countryCodeAlpha2,
+        currencyToBuy
+      )
+
+      return { providers, availableLocalProviders }
```
Comment (lisaqyli) [inline src/fiatExchanges/SelectProvider.tsx:97]: since we are refactoring this code, I find `providers` vs `availableLocalProviders` confusing.  can I understand `availableLocalProviders` as mobile money? Does `availableLocalProviders` include anything other than mobile money?

File: src/fiatExchanges/SelectProvider.tsx (hunk 3 above)
Comment (lisaqyli) [reply src/fiatExchanges/SelectProvider.tsx:225]: interesting find :)

File: src/fiatExchanges/SelectProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+}
+
+interface PaymentMethodSectionProps {
+  paymentMethod: PaymentMethod
+  cicoQuotes: CicoQuote[]
+  setNoPaymentMethods: React.Dispatch<React.SetStateAction<boolean>>
+  quoteOnPress: (cicoQuote: CicoQuote) => void
+}
+
+export function PaymentMethodSection({
```
Comment (lisaqyli) [inline src/fiatExchanges/SelectProvider.tsx:264]: i find this file gigantic, i wonder if we can  make `PaymentMethodSection` its own component

File: src/fiatExchanges/SelectProvider.tsx [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+                  <Text style={styles.category}>
+                    {paymentMethod === PaymentMethod.Card
+                      ? t('selectProviderScreen.card')
+                      : t('selectProviderScreen.bank')}
+                  </Text>
+                  <Text style={styles.fee}>
+                    {renderFeeAmount(sectionQuotes[0].quote, t('selectProviderScreen.fee'))}
+                  </Text>
+                  <Text style={styles.topInfo}>
+                    {t('selectProviderScreen.idRequired')} |{' '}
```
Comment (lisaqyli) [inline src/fiatExchanges/SelectProvider.tsx:364]: Does Ramp require ID now?

File: src/account/FiatExchange.tsx [hunk 15]
```diff
@@ -74,28 +71,28 @@ function FiatExchange() {
...
           <ListItem onPress={goToAddFunds}>
             <Text testID="addFunds" style={styles.optionTitle}>
-              {t('addFunds')}
+              {t(`fiatExchangeFlow.${FiatExchangeFlow.CashIn}.fiatExchangeTitle`)}
+            </Text>
+            <Text style={styles.optionSubtitle}>
+              {t(`fiatExchangeFlow.${FiatExchangeFlow.CashIn}.fiatExchangeSubtitle`)}
+            </Text>
+          </ListItem>
+          <ListItem onPress={goToSpend}>
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/account/FiatExchange.tsx:80]: No, that flag is just a hardcoded config anyways, is 18 months old and misleading. I'm going to deep purge it from the code.

File: src/account/FiatExchange.tsx [hunk 16]
```diff
@@ -74,28 +71,28 @@ function FiatExchange() {
...
+          <ListItem onPress={goToCashOut}>
+            <Text testID="cashOut" style={styles.optionTitle}>
+              {t(`fiatExchangeFlow.${FiatExchangeFlow.CashOut}.fiatExchangeTitle`)}
+            </Text>
+            <Text style={styles.optionSubtitle}>
+              {t(`fiatExchangeFlow.${FiatExchangeFlow.CashOut}.fiatExchangeSubtitle`)}
             </Text>
-            <Text style={styles.optionSubtitle}>{t('addFundsSubtitle')}</Text>
           </ListItem>
-          {features.SHOW_CASH_OUT ? (
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/account/FiatExchange.tsx:81]: Its a hardcoded config set to true that hasn't been touched in 2 years :) 

removed it.

File: src/fiatExchanges/FiatExchangeAmount.tsx [hunk 17]
```diff
@@ -289,9 +254,7 @@ function FiatExchangeAmount({ route }: Props) {
       >
         <View style={styles.amountInputContainer}>
           <View>
-            <Text style={styles.exchangeBodyText}>{`${t('amount')} ${
-              currency === Currency.Celo ? '(CELO)' : ''
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/FiatExchangeAmount.tsx:293]: I think this change was missed in this commit https://github.com/valora-inc/wallet/commit/7866d008145b6f35b1b47060b1c616effb6ec450

That same text is now in the header instead of here.

File: src/fiatExchanges/FiatExchangeAmount.tsx [hunk 18]
```diff
@@ -354,16 +317,14 @@ FiatExchangeAmount.navOptions = ({
...
-        title={i18n.t('addFundsHeaderWithCurrency', {
-          token: route.params.currency === Currency.Celo ? 'CELO' : route.params.currency,
-        })}
-      />
-    ) : (
-      <HeaderTitleWithBalance title={i18n.t('cashOut')} token={route.params.currency} />
-    ),
+  headerTitle: () => (
+    <HeaderTitleWithBalance
+      title={i18n.t(`fiatExchangeFlow.${route.params.flow}.exchangeAmmountTitle`, {
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/FiatExchangeAmount.tsx:322]: my most misspelled word 🤦

File: src/fiatExchanges/FiatExchangeCurrency.tsx (hunk 6 above)
Comment (jh2oman) [reply src/fiatExchanges/FiatExchangeCurrency.tsx:161]: Issue created https://app.zenhub.com/workspaces/acquisition-squad-sprint-board-6010683afabec1001a090887/issues/valora-inc/wallet/2450

File: src/fiatExchanges/utils.tsx [hunk 19]
```diff
@@ -12,6 +10,23 @@ import Logger from 'src/utils/Logger'
 
 const TAG = 'fiatExchanges:utils'
 
+export enum FiatExchangeFlow {
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/utils.tsx:13]: Good call on the nit.

Can you rephrase what you are suggesting with the second part? I don't understand. Its a bit tricky and i had trouble coming up with a good solution. The beginning of the flow can be CashIn/CashOut/Spend but the second half of the flow can only be CashIn/CashOut and i thought it would be cleaner for types if they were separate

[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/utils.tsx:13]: Oh and actually on the nit, i forgot I actually tried to do something like that and it doesn't work. Declaring something as a type has different behavior than declaring it as an enum and you can't do things like Unions with enums unfortunately

File: src/fiatExchanges/SelectProvider.tsx (hunk 12 above)
Comment (jh2oman) [reply src/fiatExchanges/SelectProvider.tsx:97]: Yea i hate the name too. changing

File: src/fiatExchanges/utils.tsx [hunk 20]
```diff
@@ -144,48 +238,24 @@ export const fetchSimplexPaymentData = async (
...
-  }
-
-  if (provider2.restricted && !provider1.restricted) {
-    return -1
-  }
-
-  const providerFee1 = getLowestFeeValueFromQuotes(provider1.quote)
-  const providerFee2 = getLowestFeeValueFromQuotes(provider2.quote)
+export const sortQuotes = ({ quote: quote1 }: CicoQuote, { quote: quote2 }: CicoQuote) => {
+  const providerFee1 = getFeeValueFromQuotes(quote1)
```
Comment (cajubelt) [inline src/fiatExchanges/utils.tsx:257]:
```suggestion
  const providerFee1 = getFeeValueFromQuotes(quote1) ?? 0
  const providerFee2 = getFeeValueFromQuotes(quote2) ?? 0
  return providerFee1 - providerFee2
```
code golf (for this whole sort callback)

File: src/fiatExchanges/utils.tsx [hunk 21]
```diff
@@ -144,48 +238,24 @@ export const fetchSimplexPaymentData = async (
...
-    return 1
-  }
-
-  if (provider2.restricted && !provider1.restricted) {
-    return -1
-  }
-
-  const providerFee1 = getLowestFeeValueFromQuotes(provider1.quote)
-  const providerFee2 = getLowestFeeValueFromQuotes(provider2.quote)
+export const sortQuotes = ({ quote: quote1 }: CicoQuote, { quote: quote2 }: CicoQuote) => {
```
Comment (cajubelt) [inline src/fiatExchanges/utils.tsx:256]: nit: could rename to make it clearer what the sort criterion is. (like `sortQuotesLowerFeeFirst`)

Comment (cajubelt) [reply src/fiatExchanges/utils.tsx:256]: or `sortQuotesByFeeAsc`

File: src/fiatExchanges/SelectProvider.tsx [hunk 22]
```diff
@@ -0,0 +1,304 @@
...
+        quoteOnPress={quoteOnPress}
+      />
+      <PaymentMethodSection
+        cicoQuotes={cicoQuotes}
+        paymentMethod={PaymentMethod.Bank}
+        setNoPaymentMethods={setNoPaymentMethods}
+        quoteOnPress={quoteOnPress}
+      />
+      <LegacyMobileMoneySection
+        provider={asyncProviders.result?.legacyMobileMoneyProviders[0]}
```
Comment (cajubelt) [inline src/fiatExchanges/SelectProvider.tsx:137]: what if there are multiple legacy mobile money providers?

**Review by cajubelt — COMMENTED** [review]
reviewed SetProvider and some deps

File: locales/base/translation.json [hunk 23]
```diff
@@ -1217,5 +1217,47 @@
...
+    "idRequired": "ID Required",
+    "numDays": "1-3 Days",
+    "oneHour": "Less Than 1 Hour",
+    "bestRate": "Best Rate",
+    "header": "Select Payment Method"
+  },
+  "fiatExchangeFlow": {
+    "CashIn": {
+      "fiatExchangeTitle": "Add Funds",
+      "fiatExchangeSubtitle": "Fund your Valora account from a bank, card or exchange.",
```
Comment (jeanregisser) [inline locales/base/translation.json:1245]: We have a globally available `{{appName}}` variable.

```suggestion
      "fiatExchangeSubtitle": "Fund your {{appName}} account from a bank, card or exchange.",
```

File: src/fiatExchanges/FiatExchangeAmount.tsx [hunk 24]
```diff
@@ -354,16 +317,14 @@ FiatExchangeAmount.navOptions = ({
...
-        title={i18n.t('addFundsHeaderWithCurrency', {
-          token: route.params.currency === Currency.Celo ? 'CELO' : route.params.currency,
-        })}
-      />
-    ) : (
-      <HeaderTitleWithBalance title={i18n.t('cashOut')} token={route.params.currency} />
-    ),
+  headerTitle: () => (
+    <HeaderTitleWithBalance
+      title={i18n.t(`fiatExchangeFlow.${route.params.flow}.exchangeAmountTitle`, {
```
Comment (jeanregisser) [inline src/fiatExchanges/FiatExchangeAmount.tsx:322]: Though this works, having a dynamic string will make it more difficult to switch to extracting i18n strings or validate keys statically. For instance https://www.i18next.com/how-to/extracting-translations#2-using-an-extraction-tool
Could you use a `switch` and static strings instead?

File: src/navigator/TopBarButton.tsx [hunk 25]
```diff
@@ -13,17 +13,28 @@ interface CommonProps {
   testID?: string
   onPress: () => void
   eventName?: AnalyticsEventType
+  eventProperties?: any
```
Comment (jeanregisser) [inline src/navigator/TopBarButton.tsx:16]: We should type this fully to avoid errors as the code evolves.
Similar to https://github.com/valora-inc/wallet/blob/2b28b691640c52bdeeb1b49a7751c5bdf3b764fe/src/analytics/ValoraAnalytics.ts#L118-L121
I'm happy to help.

File: src/analytics/Events.tsx [hunk 26]
```diff
@@ -430,6 +430,8 @@ export enum FiatExchangeEvents {
   cico_spend_select_provider_back = 'cico_spend_select_provider_back',
   cico_non_celo_exchange_send_bar_continue = 'cico_non_celo_exchange_send_bar_continue', // When send bar is tapped from cash out for cUSD & cEUR
   cico_celo_exchange_send_bar_continue = 'cico_celo_exchange_send_bar_continue', // When withdraw bar is tapped from cash out CELO
+
+  cico_select_currency_back = 'cico_select_currency_back',
```
Comment (jeanregisser) [inline src/analytics/Events.tsx:434]: nit: could you add a comment? I know we haven't been very thorough about this, but it helps tremendously when checking analytics behavior.

**Review by jeanregisser — COMMENTED** [review]
Great work!! 🎉 
Left some small comments.

File: src/fiatExchanges/SelectProvider.tsx [hunk 27]
```diff
@@ -0,0 +1,304 @@
...
+        digitalAsset={digitalAsset}
+      />
+      <ExchangesSection selectedCurrency={route.params.selectedCrypto} flow={flow} />
```
Comment (jophish) [inline src/fiatExchanges/SelectProvider.tsx:140]: Could it ever be the case that no exchanges are available?

File: src/fiatExchanges/PaymentMethodSection.tsx [hunk 28]
```diff
@@ -0,0 +1,245 @@
...
+                    {paymentMethod === PaymentMethod.Card
+                      ? t('selectProviderScreen.card')
+                      : t('selectProviderScreen.bank')}
+                  </Text>
+                  <Text style={styles.fee}>
+                    {renderFeeAmount(sectionQuotes[0].quote, t('selectProviderScreen.fee'))}
+                  </Text>
+                  <Text style={styles.topInfo}>
+                    {t('selectProviderScreen.idRequired')} |{' '}
+                    {paymentMethod === PaymentMethod.Card
```
Comment (jophish) [inline src/fiatExchanges/PaymentMethodSection.tsx:127]: This assumes that all card transfers will settle in < 1 hour, and all bank transfers will take 1-3 days, right? The  behavior here is sort of opaque, but I imagine that this will have to be generalized/refactored anyhow once FiatConnect providers are supported (in particular, since we'll be receiving this sort of information dynamically).

File: src/fiatExchanges/PaymentMethodSection.tsx [hunk 29]
```diff
@@ -0,0 +1,245 @@
...
+          <Expandable
+            arrowColor={colors.greenUI}
+            containerStyle={{
+              ...styles.expandableContainer,
+              paddingVertical: isExpandable ? (expanded ? 22 : 27) : 16,
+            }}
+            isExpandable={isExpandable}
+            isExpanded={expanded}
+          >
+            {isExpandable ? (
```
Comment (jophish) [inline src/fiatExchanges/PaymentMethodSection.tsx:93]: these nested inline ternarys are a bit tricky to follow imo

File: src/fiatExchanges/PaymentMethodSection.tsx [hunk 30]
```diff
@@ -0,0 +1,245 @@
...
+        sectionQuotes.map((cicoQuote, index) => (
+          <Touchable onPress={(quoteOnPress(cicoQuote) as unknown) as () => void}>
+            <View style={styles.expandedContainer}>
+              <View style={styles.left}>
+                <Text style={styles.expandedFee}>
+                  {renderFeeAmount(cicoQuote.quote, t('selectProviderScreen.fee'))}
+                </Text>
+                <Text style={styles.expandedInfo}>
+                  {t('selectProviderScreen.idRequired')} |{' '}
+                  {paymentMethod === PaymentMethod.Card
```
Comment (jophish) [inline src/fiatExchanges/PaymentMethodSection.tsx:156]: Looks like there's some duplicated logic for displaying settlement times between here and line 127... would be good to deduplicate if possible, though nbd

**Review by jophish — COMMENTED** [review]
This is awesome! Left a couple of comments. I'm really curious to see how this will change when we start supporting FiatConnect providers. (e.g., will we coerce all "legacy" provider quotes to look like FiatConnect quotes, vice-versa, keep them distinct?)

File: src/fiatExchanges/SelectProvider.tsx (hunk 22 above)
Comment (jh2oman) [reply src/fiatExchanges/SelectProvider.tsx:137]: Its only two and each are only available in different countries. They haven't been touched in over a year and I'm hedging my bets that we won't be adding more

File: src/fiatExchanges/FiatExchangeAmount.tsx [hunk 31]
```diff
@@ -170,40 +153,27 @@ function FiatExchangeAmount({ route }: Props) {
...
         // Rounding up to avoid decimal errors from providers. Won't be
         // necessary once we support inputting an amount in both crypto and fiat
         fiat: Math.round(localCurrencyAmountRequested.toNumber()),
       },
-      paymentMethod: route.params.paymentMethod,
     })
   }
 
   function onPressContinue() {
-    if (!route.params.isCashIn && currency === Currency.Celo) {
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/FiatExchangeAmount.tsx:195]: Safe to remove since, at minimum, external exchanges support celo cash outs

File: src/fiatExchanges/SelectProvider.tsx (hunk 27 above)
Comment (jh2oman) [reply src/fiatExchanges/SelectProvider.tsx:140]: Good question! I asked @lisaqyli if thats possible and she said there always is an option for SOME currency (usually CELO). If no exchanges are available the exchange screen actually encourages you to switch currency and try again. I think this is probably fine for now, but definitely could be optimized

File: src/fiatExchanges/SelectProvider.tsx (hunk 14 above)
Comment (lisaqyli) [reply src/fiatExchanges/SelectProvider.tsx]: might be a good idea to conditionally show `ID Required` text, I'm worried this will hurt cash in rate

File: src/navigator/types.tsx [hunk 32]
```diff
@@ -230,6 +219,14 @@ export type StackParamList = {
     selectedCountryCodeAlpha2: string
   }
   [Screens.SelectLocalCurrency]: undefined
+  [Screens.SelectProvider]: {
+    flow: CICOFlow
+    selectedCrypto: Currency
+    amount: {
+      crypto: number
+      fiat: number
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/navigator/types.tsx:227]: It looks like most of the CICO flow uses number as opposed to BigNumber. I'm going to create a separate followup ticket for going through and fixing all of that.

https://app.zenhub.com/workspaces/acquisition-squad-sprint-board-6010683afabec1001a090887/issues/valora-inc/wallet/2461

File: src/fiatExchanges/FiatExchangeAmount.tsx (hunk 24 above)
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/FiatExchangeAmount.tsx:322]: yes because its not within the react component

File: src/fiatExchanges/SelectProvider.tsx [hunk 33]
```diff
@@ -0,0 +1,304 @@
...
+        setNoPaymentMethods={setNoPaymentMethods}
+        quoteOnPress={quoteOnPress}
+      />
+      <PaymentMethodSection
+        cicoQuotes={cicoQuotes}
+        paymentMethod={PaymentMethod.Bank}
+        setNoPaymentMethods={setNoPaymentMethods}
+        quoteOnPress={quoteOnPress}
+      />
+      <LegacyMobileMoneySection
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/SelectProvider.tsx:136]: No, these legacy mobile money options are only available in two countries, so we'd essentially be always showing the limited providers text if we did that.

[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/SelectProvider.tsx:136]: And the question of how they'll interact I'm saving for later. Already enough complexity in this PR

File: src/fiatExchanges/PaymentMethodSection.tsx (hunk 29 above)
Comment (jh2oman) [reply src/fiatExchanges/PaymentMethodSection.tsx:93]: i'll clean it up a bit and separate them out into helper functions

File: src/fiatExchanges/SelectProvider.tsx (hunk 14 above)
Comment (jh2oman) [reply src/fiatExchanges/SelectProvider.tsx]: I ran it by Laura when I was working on the flow and she said its okay. The problem is our existing external providers don't provide that info so we'd have to store it somewhere and we are trying to avoid having more detailed data that can easily go stale like that per partner.

File: src/fiatExchanges/PaymentMethodSection.tsx [hunk 34]
```diff
@@ -0,0 +1,246 @@
...
+      <View style={styles.right}>
+        <Text style={styles.providerDropdown}>
+          {t('selectProviderScreen.numProviders', { count: sectionQuotes.length })}
+        </Text>
+      </View>
+    </>
+  )
+
+  const renderNonExpandableSection = () => (
+    <>
```
Comment (cajubelt) [inline src/fiatExchanges/PaymentMethodSection.tsx:72]: nit: I might add error logging if sectionQuotes had length not equal to 1, since it violates an assumption this function relies on

File: src/fiatExchanges/PaymentMethodSection.tsx [hunk 35]
```diff
@@ -0,0 +1,246 @@
...
+      <View style={styles.left}>
+        <Text style={styles.category}>
+          {paymentMethod === PaymentMethod.Card
+            ? t('selectProviderScreen.card')
+            : t('selectProviderScreen.bank')}
+        </Text>
+        {!expanded && (
+          <Text style={styles.fee}>
+            {renderFeeAmount(sectionQuotes[0].quote, t('selectProviderScreen.minFee'))}
```
Comment (cajubelt) [inline src/fiatExchanges/PaymentMethodSection.tsx:58]:
```suggestion
            {renderFeeAmount(sectionQuotes[0].quote, t('selectProviderScreen.minFee'))}  // quotes assumed to be sorted ascending by fee
```

**Review by cajubelt — COMMENTED** [review]
I noticed the test instructions are a little vague-- doesn't have to block this PR but I'd like for us to provide QA with more specific recommendations. For example, it's probably worth mentioning to them that the CICO flow from entering the provider widget onwards should not be affected (or require testing).

File: src/fiatExchanges/SelectProvider.tsx (hunk 22 above)
Comment (jh2oman) [reply src/fiatExchanges/SelectProvider.tsx:136]: Its not that simple. The mobile money section will need to be similar to the Bank/Card sections showing an expandable dropdown if there are multiple. My reasons for not doing it in this PR include:
* The PR was already getting huge, and its not something we need yet.
* The best way to do it would probably be to make [PaymentMethodSection](https://github.com/valora-inc/wallet/blob/select-provider-squash/src/fiatExchanges/PaymentMethodSection.tsx) more flexible, but I'd like to wait until after [Wallet call fetchFiatConnectProviders and display](https://app.zenhub.com/workspaces/acquisition-squad-sprint-board-6010683afabec1001a090887/issues/valora-inc/wallet/2332) before trying to reconcile legacyMobileMoney data structure with quotes
* The next mobile money provider we add will likely be fiatconnect and so that will require wallet changes anyways

I'll bring the decision to only use the first legacyMobileMoneyProvider into the component and add a comment to clarify so that its less likely that this will be overlooked and taken as a pattern

Comment (cajubelt) [reply src/fiatExchanges/SelectProvider.tsx:136]: right, the expandable thing. Ok, makes sense. If the assumption is made more explicit I think I'm fine with it

File: src/fiatExchanges/PaymentMethodSection.tsx (hunk 34 above)
Comment (jh2oman) [reply src/fiatExchanges/PaymentMethodSection.tsx:72]: I don't think so. Thats one of those logs where its programmatically impossible to reach with the current code. Those types of issues should be covered by unit tests, which this is :)

File: src/fiatExchanges/FiatExchangeCurrency.test.tsx [hunk 36]
```diff
@@ -33,4 +34,45 @@ describe('FiatExchangeCurrency', () => {
...
+    const tree = render(
+      <Provider store={createMockStore({})}>
+        <FiatExchangeCurrency {...mockScreenProps(FiatExchangeFlow.CashIn)} />
+      </Provider>
+    )
+
+    fireEvent.press(tree.getByTestId('radio/CELO'))
+    fireEvent.press(tree.getByText('next'))
+    expect(navigate).toHaveBeenCalledWith(Screens.FiatExchangeAmount, {
+      currency: 'cGLD',
```
[the remark this answers is not in the document]
Comment (lisaqyli) [reply src/fiatExchanges/FiatExchangeCurrency.test.tsx:61]: i think celo is still listed as CGLD on coinbase

File: src/fiatExchanges/FiatExchangeCurrency.tsx [hunk 37]
```diff
@@ -0,0 +1,225 @@
...
+    </SafeAreaView>
+  )
+}
+
+const styles = StyleSheet.create({
+  content: {
+    flex: 1,
+    flexDirection: 'column',
+    justifyContent: 'space-between',
+    backgroundColor: colors.gray1,
```
[the remark this answers is not in the document]
Comment (jh2oman) [reply src/fiatExchanges/FiatExchangeCurrency.tsx]: good catch, ty!

**Review by cajubelt — APPROVED** [review]
🥳

**Review by lisaqyli — APPROVED** [review]

**Comment by jh2oman** [comment]
Comment 100 🙃
